lib/using_restricted_array.rb

```diff
@@ -1,82 +1,140 @@
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:linear or O(n), n is the value_to_find of items in the array or the length of the input array
+# Space complexity: O(1)
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  len = length(array)
+  i = 0
+  while i < len
+    print array[i]
+    print " "
+    i += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    if array[i] == value_to_find
+      return true
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  largest_number = array[0]
+  length.times do |i|
+    if array[i] > largest_number
+      largest_number = array[i]
+    end
+  end
+  return largest_number
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n)
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest_number = array[0]
+  length.times do |i|
+    if smallest_number > array[i]
+      smallest_number = array[i]
+    end
+  end
+  return smallest_number
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  n = 1
+  (length / 2).times do |i|
+    temp = 0
+    temp = array[i]
+    array[i] = array[length - n]
+    array[length - n] = temp
+    n += 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log n), logarithmic
+# Space complexity: O(1)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  if length == 0
+    return false
+  end
+  low_point = 0
+  high_point = length - 1
+  while low_point < high_point
+    mid = (low_point + high_point) / 2
+    if array[mid.floor] == value_to_find
+      return true
+    elsif array[mid.floor] > value_to_find
+      high_point = mid - 1
+    elsif low_point = mid + 1
```

Comment:
I don't recommend doing assignment inside of an `if` or `elsif`. It raises alarm bells in my head because most of the time when you see this in code, it's a mistake where someone was trying to do a comparison (`==`) instead of assignment (`=`).  Just make this instead be:

```ruby
else
  low_point = mid + 1
end
```